athena/etc/console/console.8
.TH CONSOLE 8 "9 August 1991"
.ds ]W MIT Project Athena
.SH NAME
console \- X console window
.SH SYNOPSIS
.nf
console [-blink \fImilliseconds\fR] [-unmap \fIseconds\fR] [\(+-map]
        [-global] [-nosession]
.fi
.SH DESCRIPTION
The \fIconsole\fR program reads standard input and displays it
in an X window on the screen.  It is intended to be started by \fIdm\fR with
its input connected to the logical console for the workstation.  Its
default behavior is to start up unmapped (not visible on the screen),
reading its initial buffer contents from \fI/var/athena/console.log\fR. 
It will map itself (i.e. become visible on the screen) whenever
there is new input.  The user may then click its "Hide"
button to hide the window again until new input comes.  
For other available behaviors, see below.
.SH SIGNALS
\fIConsole\fR takes the following signals, which are usually sent to
it by \fIconfig_console\fR, \fIdm\fR, or \fInanny\fR:
.TP 8
.B SIGHUP
On a hangup signal, \fIconsole\fR attempts to dump the contents
of its text buffer to \fI/var/athena/console.log\fR, and then exits. (\fIDm\fR
will then usually restart it.)
.TP 8
.B SIGFPE
On an FPE signal, \fIconsole\fR zeroes its text buffer, attempts to
zero \fI/var/athena/console.log\fR, and unmaps its window.
.TP 8
.B SIGUSR1
On a USR1 signal, \fIconsole\fR maps its window.
.TP 8
.B SIGUSR2
On a USR2 signal, \fIconsole\fR unmaps its window.
.TP 8
.B SIGBUS
On a BUS signal, \fIconsole\fR dumps core.
.SH OPTIONS
\fIConsole\fR accepts a subset of the standard X Toolkit command line
options, namely: -display, -geometry, -bg, -background, -bd,
-bordercolor, -bw, -borderwidth, -fg, -foreground, -fn, -font, -name,
\(+-rv, \(+-reverse, \(+-iconic, -icongeometry, -xrm. See X(1) for
details on these options.

\fIConsole\fR accepts five additional options:
.TP 8
.B \-blink \fImilliseconds\fR
This option specifies the blink frequency in milliseconds of
\fIconsole\fR's icon. If \fIconsole\fR is iconified, and its
icon window is visible, it will blink its icon at this frequency
when there is new output to be displayed, rather than mapping
itself. The default frequency is 1000 milliseconds (1 second).
The minimum frequency is 100 milliseconds.
.TP 8
.B \-unmap \fIseconds\fR
If this option is specified and nonzero, \fIconsole\fR will
automatically unmap itself after this many seconds of no output.
"-unmap 10" for example will cause the window to disappear
by itself 10 seconds after it appears with new input, saving the user 
from continually having to click the "Hide" button while working.
.TP 8
.B \(+-map
This option specifies whether the console window should start up mapped
(visible).  -map causes the window to start mapped, +map causes
\fIconsole\fR to wait until there is input before mapping the window on 
the screen.  Starting unmapped is the default.
.TP 8
.B \-global
When specified, this option causes \fIconsole\fR to place a
_SGI_DESKS_HINTS property on its window, with the contents
_SGI_DESKS_ALWAYS_GLOBAL. This is only useful when the console
window is running with SGI's 4Dwm window manager, where it will
cause the console to appear on all desks (if desks are in use).
.TP 8
.B \-nosession
This option causes \fIconsole\fR not to place a WM_COMMAND property
on its window. This is needed to prevent session managers
(such as SGI's 4Dwm window manager) from attempting to start up \fIconsole\fR
themselves when it is being used as the system console window.
.SH X RESOURCES
Since \fIconsole\fR is normally started up before the user's resources
have been read into the server, it does not have the chance to read the
user's preferences.  If the user wishes \fIconsole\fR to use her
resources, she may run the program \fIconfig_console\fR. See the man
page for \fIconfig_console\fR for more information.

\fIConsole\fR understands the following resources:
.TP 8
.B *foreground
This is the same as the -fg option.
.TP 8
.B *background
This is the same as the -bg option.
.TP 8
.B *reverseVideo
This is the same as the \(+-rv option. Specify True or False.
.TP 8
.B *font
This is the same as the -fn option.
.TP 8
.B *borderWidth
This is the same as the -bw option.
.TP 8
.B *borderColor
This is the same as the -bd option.
.TP 8
.B Console*window.geometry
This is the same as the -geometry option.
.TP 8
.B Console*iconWindow.geometry
This is the same as the -icongeometry option.
.TP 8
.B Console*frequency
This is the same as the -blink option.
.TP 8
.B Console*autoUnmap
This is the same as the -unmap option.
.TP 8
.B Console*window.mapped
This is the same as \(+-map. Specify True or False.
.TP 8
.B Console*window.iconic
This is the same as \(+-iconic. Specify True or False.
.TP 8
.B Console*window.global
This is the same as \-global. Specify True or False.
.TP 8
.B Console*window.showCommand
This is the same as \-nosession (which sets this to False).
.SH SELECTING TEXT
Console provides the ability to select text from the window for "pasting"
into other applications (xterm, emacs, etc.).  The method for "cutting"
text closely emulates the behavior of xterm.
.PP
Mouse-buttons one and three are used to select text.  Move the
mouse-pointer to the beginning of the region you wish to select.  Press
and hold down button one, and drag the mouse to the end of the region
you wish to select, and release button one.  The current selection will
be highlighted as you drag the mouse.
.PP
Alternatively, you can set the beginning of the region by clicking
mouse-button one and then using mouse-button three to extend the
selection.
.PP
Mouse-button three "extends" the current selection.  The selection is
expanded or contracted from the edge closest to the mouse-pointer.
Holding down the mouse-button while extending the selection allows you
to "drag" out an extension.
.PP
Text can be selected in one of five "modes".  Clicking the mouse button
rapidly cycles through the modes.  Double-clicking selects by words.
Triple-clicking selects by lines.  Quad-clicking selects by paragraph,
where a paragraph break is defined by a blank line.  Quint-clicking
selects the entire text of the console.
.SH USING THE SCROLLBAR
The scrollbar thumb (the white rectangle over a gray background)
displays the range of text visible in relation to the total amount of
text in the console.
.PP
Clicking button one in the scroll-region scrolls the text up by the
number of lines above the pointer.  That is, the closer the the pointer
is to the bottom of the scroll region, the more the text will scroll.
.PP
Clicking button three in the scroll-region scrolls the text down by the
number of lines above the pointer, in the same way that button one
scrolls up.
.PP
Clicking button two in the scroll-region scrolls the console to the place
in the text that corresponds to the pointer's position in the scrollbar.
The thumb may be "dragged" by pressing mouse-button two in the
scroll-region and moving the mouse while the mouse-button is
depressed.
.SH FILES
.PP
/var/athena/console.log    console log file
.br
/etc/athena/login/Console  application defaults
.br
/var/athena/console.pid    pid of console, if started by \fIdm\fR
.SH "SEE ALSO"
X(1), config_console(1), show_console(1), hide_console(1), dm(8), nanny(8),
xrdb(1)
.SH "BUGS"
\fIConsole\fR may sometimes unmap its window (if -unmap is specified)
while you are reading it.  Moving the scrollbar before the specified timeout
will prevent this.

\fIConsole\fR should provide more control over its mapping and blinking
icon abilities.

Variable-width fonts are not supported.

Quad-clicking effectively selects the entire console text, since it all
appears to be one paragraph; the time-stamping at the beginning of each
line leaves no blank lines, thus the entire text is one paragraph.
.SH AUTHORS
Craig Fields and Chris VanHaren, MIT Project Athena
.br
Copyright (c) 1990-1995 Massachusetts Institute of Technology


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/etc/console/console.c
/* Copyright 1990, 1991, 1998 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

static const char rcsid[] = "$Id: console.c,v 1.1 1998-12-17 16:26:02 ghudson Exp $";

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/time.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <errno.h>
#include <signal.h>
#include <syslog.h>
#include <Jets/Jets.h>
#include <Jets/Window.h>
#include <Jets/Button.h>
#include <Jets/Label.h>
#include <Jets/Form.h>
#include <Jets/ScrollBar.h>
#include <Jets/TextDisplay.h>
#include <Jets/Arrow.h>
#include <X11/Xresource.h>

#ifndef CONSOLEDEFAULTS
#define CONSOLEDEFAULTS "/etc/athena/login/Console"
#endif

#define XENV "XENVIRONMENT"
#define CONSOLEFILE "/var/athena/console.log"

extern int DEBUG;

static XrmOptionDescRec opTable[] = {
{"+rv",         "*reverseVideo", XrmoptionNoArg,        (caddr_t) "off"},
{"-background", "*background",  XrmoptionSepArg,        (caddr_t) NULL},
{"-bd",         "*window.borderColor", XrmoptionSepArg,	(caddr_t) NULL},
{"-bg",         "*background",  XrmoptionSepArg,        (caddr_t) NULL},
{"-bordercolor","*borderColor", XrmoptionSepArg,        (caddr_t) NULL},
{"-borderwidth","*window.borderWidth", XrmoptionSepArg, (caddr_t) NULL},
{"-bw",         "*window.borderWidth", XrmoptionSepArg, (caddr_t) NULL},
{"-display",    ".display",     XrmoptionSepArg,        (caddr_t) NULL},
{"-fg",         "*foreground",  XrmoptionSepArg,        (caddr_t) NULL},
{"-fn",         "*font",        XrmoptionSepArg,        (caddr_t) NULL},
{"-font",       "*font",        XrmoptionSepArg,        (caddr_t) NULL},
{"-foreground", "*foreground",  XrmoptionSepArg,        (caddr_t) NULL},
{"-geometry",   "*window.geometry", XrmoptionSepArg,    (caddr_t) NULL},
{"-icongeometry", "*iconWindow.geometry", XrmoptionSepArg, (caddr_t) NULL},
{"-name",       ".name",        XrmoptionSepArg,        (caddr_t) NULL},
{"-reverse",    "*reverseVideo", XrmoptionNoArg,        (caddr_t) "on"},
{"-rv",         "*reverseVideo", XrmoptionNoArg,        (caddr_t) "on"},
{"-xrm",        NULL,           XrmoptionResArg,        (caddr_t) NULL},
{"-appdefs",	".appDefs",	XrmoptionSepArg,	(caddr_t) NULL},
{"-f",		".appDefs",	XrmoptionSepArg,	(caddr_t) NULL},
{"-blink",      ".frequency",   XrmoptionSepArg,        (caddr_t) NULL},
{"-unmap",      ".autoUnmap",   XrmoptionSepArg,        (caddr_t) NULL},
{"-map",        "*window.mapped", XrmoptionNoArg, 	(caddr_t) "True"},
{"+map",        "*window.mapped", XrmoptionNoArg, 	(caddr_t) "False"},
{"-input",	".input",	XrmoptionSepArg,	(caddr_t) NULL},
{"-inputfd",	".inputfd",	XrmoptionSepArg,	(caddr_t) NULL},
{"-nostdin",	".nostdin",	XrmoptionNoArg,		(caddr_t) "True"},
{"-iconic",	"*window.iconic", XrmoptionNoArg,	(caddr_t) "True"},
{"+iconic",	"*window.iconic", XrmoptionNoArg,	(caddr_t) "False"},
{"-notimestamp",".timestamp",	XrmoptionNoArg,		(caddr_t) "False"},
{"-timestamp",	".timestamp",	XrmoptionNoArg,		(caddr_t) "True"},
{"-hidelabel",	"*hideLabel.label", XrmoptionSepArg,	(caddr_t) NULL},
{"-title",	"*title.label",	XrmoptionSepArg,	(caddr_t) NULL},
{"-hideproc",	"*hideButton.activateProc", XrmoptionSepArg, (caddr_t) NULL},
{"-cc",		"*textDisplay.charClass", XrmoptionSepArg, (caddr_t) NULL},
{"-titlebar",	"*window.title",XrmoptionSepArg,	(caddr_t) NULL},
{"-nosession",	"*showCommand",	XrmoptionNoArg,		(caddr_t) "false"},
{"-global",	"*global",	XrmoptionNoArg,		(caddr_t) "true"},
{"-autoscroll",	".autoscroll",	XrmoptionNoArg,		(caddr_t) "True"},
{"-noautoscroll",".autoscroll",	XrmoptionNoArg,		(caddr_t) "False"},
{"-file",	".file",	XrmoptionSepArg,	(caddr_t) ""},
{"-debug",	".debug",	XrmoptionNoArg,		(caddr_t) "true"},
};

typedef struct _MyResources
{
  int frequency, autoUnmap;
  char *input;
  int inputfd;
  Boolean timestamp, autoscroll, nostdin;
  char *file;
  Boolean debug;
} MyResources;

typedef struct _MyResources *MyResourcesPtr;

MyResources parms;

#define offset(field) XjOffset(MyResourcesPtr,field)

static XjResource appResources[] =
{
  { "frequency", "Frequency", XjRInt, sizeof(int),
      offset(frequency), XjRInt, (caddr_t)1000 },
  { "autoUnmap", "AutoUnmap", XjRInt, sizeof(int),
      offset(autoUnmap), XjRInt, (caddr_t)0 },
  { "input", "Input", XjRString, sizeof(char *),
      offset(input), XjRString, (caddr_t)"" },
  { "inputfd", "Inputfd", XjRInt, sizeof(int),
      offset(inputfd), XjRInt, (caddr_t)0 },
  { "nostdin", "Nostdin", XjRBoolean, sizeof(Boolean),
      offset(nostdin), XjRBoolean, (caddr_t) False },
  { "timestamp", "Timestamp", XjRBoolean, sizeof(Boolean),
      offset(timestamp), XjRBoolean, (caddr_t) True },
  { "autoscroll", "Autoscroll", XjRBoolean, sizeof(Boolean),
      offset(autoscroll), XjRBoolean, (caddr_t) True },
  { "file", "File", XjRString, sizeof(char *),
      offset(file), XjRString, (caddr_t)CONSOLEFILE },
  { "debug", "Debug", XjRBoolean, sizeof(Boolean),
      offset(debug), XjRBoolean, (caddr_t) False },
};

#undef offset

#define INPUTSIZE 1024
char inputbuf[INPUTSIZE];

#define BUFSIZE 16384 /* must be at least two times as big as INPUTSIZE */
char *text;
int length;


#define HUP 0
#define FPE 1
#define USR1 2
#define USR2 3
#define NUMSIGS 4
int sigflags[NUMSIGS];
int val = 0;
Jet root;
WindowJet iconWindow;
LabelJet icon;
WindowJet win;
ScrollBarJet sj;
TextDisplayJet tj;
int inverted = False, blinking = False, timerid;
int unmaptimerid;


static int loadFile(filename, info, max)
     char *filename, *info;
     int max;
{
    int fd, size;
    struct stat buf;
    int num;
    char errtext[100];
    char *errfmt = "error %d %s `%s'";

    if (-1 == (fd = open(filename, O_RDONLY, 0)))
      {
	sprintf(errtext, errfmt, errno, "opening", filename);
	XjWarning(errtext);
	return 0;
      }

    if (-1 == fstat(fd, &buf)) /* could only return EIO */
      {
	sprintf(errtext, errfmt, errno, "in fstat for", filename);
	XjWarning(errtext);
	close(fd);
	return 0;
      }

    size = (int)buf.st_size;

    if (-1 == (num = read(fd, info, MIN(size, max - 1))))
      {
	sprintf(errtext, errfmt, errno, "reading", filename);
	XjWarning(errtext);
	close(fd);
	return 0;
      }
    close(fd);

    info[num] = '\0';
    return num;
}


static void saveFile(filename, info, length)
     char *filename, *info;
     int length;
{
  int fd;
  char errtext[100];
  char *errfmt = "error %d %s `%s'";

  if (-1 == (fd = open(filename, O_TRUNC | O_CREAT | O_WRONLY, 0644)))
    {
      sprintf(errtext, errfmt, errno, "opening", filename);
      XjWarning(errtext);
      return;
    }

  if (length != write(fd, info, length))
    {
      sprintf(errtext, errfmt, errno, "writing", filename);
      XjWarning(errtext);
      close(fd); /* I doubt it. */
      return;
    }

  close(fd);
}


int hide(fromJet, what, data)
     Jet fromJet;
     char *what;
     caddr_t data;
     /*ARGSUSED*/
{
  UnmapWindow(win);
  return 0;
}


void resetunmaptimer()
{
  if (parms.autoUnmap != 0)
    {
      if (unmaptimerid != 0)
	(void)XjRemoveWakeup(unmaptimerid);
      unmaptimerid = XjAddWakeup(hide, 0, 1000 * parms.autoUnmap);
    }
}


int scroll(fromJet, increment, data)
     Jet fromJet;
     int increment;
     caddr_t data;
     /*ARGSUSED*/
{
  int cl, vl;

  resetunmaptimer();
  cl = CountLines(tj);
  vl = VisibleLines(tj);
  val = GetScrollBarValue(sj);
  if ((increment > 0  &&  val < cl - vl)
      ||  (increment < 0   &&  val > 0))
    {
      val += increment;
      SetScrollBar(sj, 0, MAX(0, cl - 1), vl, val);
      SetLine(tj, val);
    }
  return 0;
}


int newvalue(fromJet, nothing, data)
     Jet fromJet;
     int nothing;
     caddr_t data;
     /*ARGSUSED*/
{
  resetunmaptimer();
  SetLine(tj, val = GetScrollBarValue((ScrollBarJet) fromJet));
  return 0;
}


int textscroll(fromJet, nothing, data)
     Jet fromJet;
     int nothing;
     caddr_t data;
     /*ARGSUSED*/
{
  int cl, vl;

  resetunmaptimer();
  cl = CountLines(tj);
  vl = VisibleLines(tj);
  val = TopLine(tj);
  SetScrollBar(sj, 0, MAX(0, cl - 1), vl, val);
  return 0;
}


int textresize(fromJet, nothing, data)
     Jet fromJet;
     int nothing;
     caddr_t data;
     /*ARGSUSED*/
{
  int cl, vl;

  resetunmaptimer();
  cl = CountLines(tj);
  vl = VisibleLines(tj);

  if ((val != 0) && (val > cl - vl))
    {
      if (parms.autoscroll)
	{
	  val = MAX(0, cl - vl);

	  SetLine(tj, val);
	}
    }

  SetScrollBar(sj, 0, MAX(0, cl - 1), vl, val);
  return 0;
}


int delete(fromJet, what, data)
     Jet fromJet;
     char *what;
     caddr_t data;
     /*ARGSUSED*/
{
  XjDestroyJet(fromJet);
  XCloseDisplay(root->core.display);
  XjExit(0);
  return 0;				/* For linting... */
}


void blinkoff()
{
  if (blinking)
    {
      if (inverted)
	{
	  inverted = !inverted;
	  setIcon(icon, Center, inverted);
	}

      (void)XjRemoveWakeup(timerid);
      blinking = False;
    }
}


int mapnotify(fromJet, what, data)
     Jet fromJet;
     char *what;
     caddr_t data;
     /*ARGSUSED*/
{
  blinkoff();
  return 0;
}


void OurMapWindow(who)
     WindowJet who;
{
  blinkoff();
  resetunmaptimer();
  MapWindow(who, True);
}


int blink(info, id)
     int info, id;
     /*ARGSUSED*/
{
  inverted = !inverted;
  setIcon(icon, Center, inverted);
  timerid = XjAddWakeup(blink, 0, MAX(100, parms.frequency));
  return 0;
}


int appendToBuffer(what, howmuch)
     char *what;
     int howmuch;
{
  int moved = 0;

  if (howmuch == 0)		/* if nothing to copy, return */
    return 0;

  /*
   * If there's not enough room to copy in whatever was asked for, then
   * chop off the beginning of the text, using bcopy...
   */
  if (BUFSIZE - length <= howmuch)
    {
      char *c;

      c = text + MIN(INPUTSIZE, length);
      while ((c < text + length) && *c != '\0' && *c != '\n')
	c++;
      if (*c != '\n')		/* if this is a looooong line, screw it. */
	c = text + MIN(INPUTSIZE, length);
      else
	c++;

      moved = c - text;
      length -= moved;

      memmove(text, c, length);
    }

  memcpy(text + length, what, howmuch);
  length += howmuch;
  text[length] = '\0';		/* Null terminate the string... */
				/* WARNING:  This can overflow!  Why? */
  return moved;
}


int input(fd, pfd)
     int fd;
     int *pfd;
     /*ARGSUSED*/
{
  int redo = 0;
  int cl, vl, l;
  char ctrl[2];
  int i = 0, j, err;
  static int eol = 1;

  ctrl[0] = '^';
  l = read(*pfd, inputbuf, INPUTSIZE);

  if (l < 0 && errno != EIO && errno != EINTR)
    {
      err = errno;
      syslog(LOG_ERR, "Error reading from fd %d: %m", *pfd);
      XjReadCallback((XjCallbackProc)NULL, *pfd, (caddr_t) pfd);
      sprintf(inputbuf, "Error reading from fd %d: %s\n", *pfd,
	      strerror(err));
      l = strlen(inputbuf);
    }

  if (l > 0)
    {
      while (i < l)
	{
	  if (eol && parms.timestamp)
	    {
	      struct timeval now;
	      time_t timet;
	      char *timestr;

	      gettimeofday(&now, NULL);
	      timet = (time_t) now.tv_sec;
	      timestr = (char *) ctime(&timet);
	      redo += appendToBuffer(&timestr[11], 5);
	      redo += appendToBuffer(" ", 1);
	    }

	  eol = 0;
	  j = i;
	  while (j < l && !eol)
	    {
	      if (iscntrl(inputbuf[j]))
		{
		  redo += appendToBuffer(&inputbuf[i], j - i);
		  i = j + 1;

		  switch(inputbuf[j])
		    {
		    case 7: /* bell compression option... */
		      XBell(root->core.display, 0);
		    case 0:
		    case 13:
		      break;
		    case '\n':
		      eol = 1;
		    case '\t':
		      redo += appendToBuffer(&inputbuf[j], 1);
		      break;
		    default:
		      ctrl[1] = inputbuf[j] | 64;
		      redo += appendToBuffer(ctrl, 2);
		      break;
		    }
		}
	      j++;
	    }
	  redo += appendToBuffer(&inputbuf[i], j - i);
	  i = j;
	}
      if (!redo)
	AddText(tj);		/* ??? */
      else
	MoveText(tj, text, redo);

      cl = CountLines(tj);
      vl = VisibleLines(tj);

      if (parms.autoscroll)
	{
	  val = MAX(0, cl - vl);

	  SetScrollBar(sj, 0, MAX(cl - 1, 0), vl, val);
	  SetLine(tj, val);
	}
      
      if (!WindowMapped((WindowJet)win))
	{
	  if (!WindowMapped(iconWindow))
	    OurMapWindow((WindowJet)win); /* we're hidden, not iconified */
	  else
	    {
	      if (!blinking)
		{
		  blinking = True;
		  blink(0, 0);
		}
	    }
	}
      else
	{
	  resetunmaptimer();
	  if (WindowVisibility(win) != VisibilityUnobscured)
	    XRaiseWindow(win->core.display, win->core.window); /* ICCCM ok */
	}
    }
  return 0;
}

void sighandler(sig)
     int sig;
{
  switch(sig)
    {
    case SIGHUP:
      sigflags[HUP] = 1;
      break;
    case SIGFPE:
      sigflags[FPE] = 1;
      break;
    case SIGUSR1:
      sigflags[USR1] = 1;
      break;
    case SIGUSR2:
      sigflags[USR2] = 1;
      break;
    }
  return;
}


XjCallbackRec callbacks[] =
{
  { "newvalue", newvalue },
  { "textresize", textresize },
  { "textscroll", textscroll },
  { "delete", delete },
  { "hide", hide },
  { "mapnotify", mapnotify },
  { "scroll", scroll },
};


static int checkSignals()
{
  int sigs;
  int cl, vl;
  int ret_code=0;

  do
    {
      sigs = 0;

      if (sigflags[FPE])
	{
	  sigs++;
	  sigflags[FPE] = 0;
	  text[0] = '\0';
	  length = 0;
	  saveFile(parms.file, text, length);

	  blinkoff();
	  if (WindowMapped(win))
	    UnmapWindow(win);

	  SetText(tj, text);

	  cl = CountLines(tj);
	  vl = VisibleLines(tj);

	  val = MAX(0, cl - vl);

	  SetScrollBar(sj, 0, MAX(cl - 1, 0), vl, val);
	  SetLine(tj, val);
	  XFlush(tj->core.display);
	}

      if (sigflags[USR1])
	{
	  sigs++;
	  sigflags[USR1] = 0;
	  if (!WindowMapped((WindowJet)win))
	    {
	      OurMapWindow((WindowJet)win);
	      XFlush(win->core.display);
	    }
	  else
	    if (WindowVisibility(win) != VisibilityUnobscured)
	      {
		XRaiseWindow(win->core.display, win->core.window); /* ICCCM ok */
		XFlush(win->core.display);
	      }
	}

      if (sigflags[USR2])
	{
	  sigs++;
	  sigflags[USR2] = 0;
	  if (WindowMapped(win))
	    {
	      UnmapWindow(win);
	      XFlush(win->core.display);
	    }
	}

      if (sigflags[HUP])
	{
	  sigs++;
	  sigflags[HUP] = 0;
	  saveFile(parms.file, text, length);
	  XCloseDisplay(root->core.display);
	  XjExit(0);
	}
      ret_code += sigs;
    }
  while(sigs);
  return ret_code;
}


fatal(display)
     Display *display;
     /*ARGSUSED*/
{
  XjExit(1);
}


main(argc, argv)
     int argc;
     char **argv;
{
  Jet w, b, f;
  int cl, vl, i;
  int zero = 0;
  int auxinput = -1;
  int size=0;
  char *envvar;
  struct stat buf;
  struct sigaction act;
  sigemptyset(&act.sa_mask);
  act.sa_flags = 0;

  openlog("console", 0, LOG_USER);

  for (i = 0; i < NUMSIGS; i++)
    sigflags[i] = 0;

  act.sa_handler= sighandler;
  (void) sigaction(SIGHUP, &act, NULL);
  (void) sigaction(SIGFPE, &act, NULL);
  (void) sigaction(SIGUSR1, &act, NULL);
  (void) sigaction(SIGUSR2, &act, NULL);

  if (getenv(XENV) == NULL)
    {
      envvar = malloc(strlen(XENV) + strlen(CONSOLEDEFAULTS) + 2);
      if (!envvar)
	{
	  fprintf(stderr, "malloc failed!  Aborting.\n");
	  exit(1);
	}
      sprintf(envvar, "%s=%s", XENV, CONSOLEDEFAULTS);
      putenv(envvar);
    }

  root = XjCreateRoot(&argc, argv, "Console", NULL,
		      opTable, XjNumber(opTable));

  (void) XSetIOErrorHandler(fatal);

  XjLoadFromResources(NULL,
		      NULL,
		      programClass,
		      programName,
		      appResources,
		      XjNumber(appResources),
		      (caddr_t) &parms);

  DEBUG = parms.debug;

  XjRegisterCallbacks(callbacks, XjNumber(callbacks));

  iconWindow = (WindowJet) XjVaCreateJet("iconWindow",
					 windowJetClass, root, NULL, NULL);
  icon = (LabelJet) XjVaCreateJet("icon",
				  labelJetClass, iconWindow, NULL, NULL);
  XjRealizeJet(root);

  win = (WindowJet) XjVaCreateJet("window", windowJetClass, root,
				  XjNiconWindow, iconWindow,
				  NULL, NULL);
  f = XjVaCreateJet("form", formJetClass, win, NULL, NULL);

  /*
   * Button
   */
  w = XjVaCreateJet("hide", windowJetClass, f, NULL, NULL);
  b = XjVaCreateJet("hideButton", buttonJetClass, w, NULL, NULL);
  (void) XjVaCreateJet("hideLabel", labelJetClass, b, NULL, NULL);

  /*
   * Arrow buttons
   */
  w = XjVaCreateJet("scrollup", windowJetClass, f, NULL, NULL);
  b = XjVaCreateJet("scrollupButton", buttonJetClass, w, NULL, NULL);
  (void) XjVaCreateJet("scrollupArrow", arrowJetClass, b, NULL, NULL);

  w = XjVaCreateJet("scrolldown", windowJetClass, f, NULL, NULL);
  b = XjVaCreateJet("scrolldownButton", buttonJetClass, w, NULL, NULL);
  (void) XjVaCreateJet("scrolldownArrow", arrowJetClass, b, NULL, NULL);

  /*
   * Label
   */
  (void) XjVaCreateJet("title", labelJetClass, f, NULL, NULL);

  /*
   * Scrollbar
   */
  w = XjVaCreateJet("scrollBarWindow", windowJetClass, f, NULL, NULL);
  sj = (ScrollBarJet) XjVaCreateJet("scrollBar",
				    scrollBarJetClass, w, NULL, NULL);

  /*
   * TextDisplay
   */
  w = XjVaCreateJet("textDisplayWindow", windowJetClass, f, NULL, NULL);
  tj = (TextDisplayJet) XjVaCreateJet("textDisplay",
				      textDisplayJetClass, w, NULL, NULL);


  XjRealizeJet((Jet) XjParent(f));


  if (! stat(parms.file, &buf))	/* If there was an error stat'ing the */
    size = (int)buf.st_size;	/* file, we will find out soon enough */
				/* when we call loadFile below.  No need */
				/* to display an error msg here. */

  size = (size > BUFSIZE) ? size : BUFSIZE;
  text = XjMalloc(size);
  text[0] = '\0';
  length = 0;
  length = loadFile(parms.file, text, size);

  if (parms.input && *parms.input)
    {
      auxinput = open(parms.input, O_RDONLY, 0);
      if (auxinput == -1)
	{
#define CANT_OPEN "console: can't open input\n"
	  appendToBuffer(CANT_OPEN, sizeof(CANT_OPEN)-1);
	}
    }
  SetText(tj, text);

  cl = CountLines(tj);
  vl = VisibleLines(tj);

  if (parms.autoscroll)
    {
      val = MAX(0, cl - vl);

      SetScrollBar(sj, 0, MAX(cl - 1, 0), vl, val);
      SetLine(tj, val);
    }

  if (parms.nostdin == False)
    XjReadCallback((XjCallbackProc)input, zero, (caddr_t) &zero);

  if (auxinput != -1)
    XjReadCallback((XjCallbackProc)input, auxinput, (caddr_t) &auxinput);

  if (parms.inputfd != 0)
    XjReadCallback((XjCallbackProc)input, parms.inputfd,
		   (caddr_t) &parms.inputfd);

  XjSetSignalChecker(checkSignals);

  XjEventLoop(root);
}
